### ISSUE TITLE
Replace MongoEngine

### ISSUE THREAD

**Issue by mgoffin** [crits/crits#854]
Since starting use of it within the project, the featureset and performance have gone downhill. It's better we replace this with something of our own. More detailed information about why things have gone downhill can be found here:

https://github.com/marrow/contentment/issues/12

My suggestion for replacement is to redo the backend code to support interchangeable databases. Configure which database you want to use and the classes should handle doing the right thing. This will allow those folks on MongoDB to stay, and those folks who want to be on ElasticSearch to do so.

This would impact the API as well which should be cleaned up anyways given how slow it is with the multitude of layers you have to pass through to do stuff.

**Comment by mgoffin** [comment]
Thinking through this I had some ideas. The first one being the class everything inherits to gain its ability to perform database operations. I was thinking we would setup `settings.py` to allow for people to choose between the databases we have support for, so something like:

```python
DATABASE_TYPE='MongoDB'
#DATABASE_TYPE='ElasticSearch'
```

We would then have a `backend` folder which would contain a file for each backend database we wish to support. A file `crits_mongodb.py` might contain:

```python
class CRITsMongo(object):
    # c will be set to the pymongo cursor (or maybe a list?) that is returned for iteration 
    c = None
    # l will be the length of the cursor returned
    l = 0
    # The collection to query which will be overwritten by whatever class inherits this one
    coll = None

    def __init__(self, c=None):
        if c:
            self.c = c
        ....
    def objects(self):
        """ Query for content in the database""""
        ....
    def save(self):
        """ Iterate over "c" and save the content back to the database """
        ....
    def delete(self):
        """ Iterate over "c" and delete the content in the database """
        ....
```

And we would have the main class be something like:

```python
from django.conf import settings

from backend.crits_mongodb import CRITsMongo
from backend.crits_elasticsearch import CRITsElasticSearch

if settings.DATABASE_TYPE == 'MongoDB':
    backend_db = CRITsMongo
elif settings.DATABASE_TYPE == 'ElasticSearch':
    backend_db = CRITsElasticSearch

class CRITsDB(backend_db):
     def __init__(self):
        ....
```

This would allow us to create a modular backend for supporting more than one database, but still provide similar functionality for working with the data once it is in python-land. My theory for setting `c` for the `CRITsMongo` class is so you can do something like this (let's assume `Sample` inherits `CRITsDB`):

```python
from crits.samples.sample import Sample

samples = Sample.objects(query={'campaigns.name': 'foo'}, filters={'limit': 1})
for s in samples.c:
    # Set the description to "foo"
    s['description'] = 'foo'

# loop over all of the documents in "c" (which would be one) and save it back to the database
samples.save()
```

or

```python
from crits.samples.sample import Sample

samples = Sample.objects(query={'campaigns.name': 'foo'})
for s in samples.c:
    # Add 'foo' as a bucket list item to each sample returned in the query
    s['bucket_list'].append('foo')

# Save all of the documents back to the database
samples.save()
```

There's probably better/cleaner ways to do the above that make it more developer-friendly, and of course each TLO class will still have methods that can do the dirty work and do it properly so users can't screw up. They would also have attributes like the collection/table/index they are associated with, `jtable` options, etc. We can also leverage the `vocabulary` section of CRITs to build classes that contain all of the fields we support so people don't have to worry about fat-fingering them.

Another thing would be deprecating support for "unsupported_attrs". Now that we are ditching MongoEngine, we can just let people add what they want to the database and it shouldn't hurt anything if we are dealing with `dict` objects again.

**Comment by mgoffin** [comment]
Another thing of note is the `filedata` fields that we have. MongoEngine would normally turn that into a complex object even though in the database it's just an `ObjectId`.

We will need to adjust our code such that we add the file to GridFS manually, set the `ObjectId` of that GridFS document to `filedata`, and maybe write a method to make that easier as well as reading the file back out.

**Comment by packet-rat** [comment]
@mgoffin :  Excellent discussion/concepts Mike!

(1) Very interested in contributing (large scale data set testing at least).
(2) Will this ElasticSearch development effort be managed in a separate branch or in Master?

**Comment by mgoffin** [comment]
@packet-rat 

1) I'll probably start this in a branch and get the core piece of it functional without doing any replacing of MongoEngine. It'll allow for testing to see if it works, if it's more performant, and if it is easy for developers to use.

2) The ElasticSearch stuff will probably happen simultaneously in whatever branch I do the initial work in. My hope is that by developing both MongoDB and ElasticSearch backends at the same time, it'll ensure that I keep the code honest when it comes to not being db-specific.

**Comment by ckane** [comment]
@unixfreak0037

**Comment by ckane** [comment]
So I've been pondering on this ever since my crits_services presentation. I was kinda thinking along these lines:
1. Abstract away (or supplement) the mongoengine parts of the data model such that new database "drivers" can be created that provide certain interfaces and expose the data model in standard ways
2. Rewrite of the CRITs API (new v2 API) that better and more comprehensively provides access to the underlying data, still in a controlled manner
3. Rewrite of the front-end to utilize the API resources for data operations, rather than a separate ad-hoc "REST-like" interface

Was also thinking that it would be good, at some point, to fold/flatten the objects into a single "TLO" model, since app-wide unique primary keys already distinguish them, and then the compartments of CRITs that are currently delegated to different "collections" in MongoDB could instead be filters that each only show the TLO's that match their individual document/column schema for the metadata they're organizing/reporting to the client. It's kind of a flip-on-the-head of the original design, but I think it removes some rigidity and also makes the application more compatible with "BigTable" DB's like Dynamo, HBase, etc...

**Comment by mgoffin** [comment]
@ckane I suspect what I am dealing with will cover your first issue. I agree that we should move to making the core implementation a RESTful API that any interface can be built upon, and provide an interface that uses it so people can get up and running immediately.

However for this issue, I would assert that most of your second and all of your third issue are out-of-scope. The API will be affected mainly because mongoengine is going away and tastypie-mongoengine will have to go away. This will be a complete API break as the current API implementation is very MongoDB-focused and won't translate well to other backend systems.

Having all of the data in one collection (or index, or whatever it is your backend database uses) has its pros and cons. I would say, however, that the cons outweight the pros. Take MongoDB and ElasticSearch for example:

- MongoDB indexing would be impossible to do without using sparse indexes for fields that exist in some documents and not in others. We've moved away from sparse indexes due to the issue of sorting on fields with a sparse index resulting in missing content.
- ElasticSearch templates are a huge tool that we would want to look at leveraging and by jamming lots of mismatching documents into the same index you sort of make this a living nightmare to template.

If you feel it would be a benefit to do this, I'd say it would also belong in another conversation and fleshed out to see if it would have an impact on this initial cut at removing mongoengine (keep in mind that this is a first step towards allowing us to do other things which could in the future include collapsing documents into a single collection or the like).

**Comment by ckane** [comment]
No, I agree - I read the discussion above as basically attempting to abstract it enough so that ElasticSearch (or similar) could be used as a back-end. A lot of my other comments above are mainly "Things I've been pondering" which kinda got revived in my memory when I saw this discussion pop up.

**Comment by mgoffin** [comment]
Yep! My goal is to ensure that data presented to the developer after it is queried from whatever database will look identical, allowing us to do the same thing to it, and the act of saving back will DTRT for whatever database you want. I suspect, having not a lot of experience with ES that it means MongoDB collection == ElasticSearch Index, but I'll be working through that in design and seeing what people think.

**Comment by ckane** [comment]
I'd recommend reading this: https://www.elastic.co/blog/index-vs-type

What ES considers "type" is probably more akin to what I see as the IPs-Domains-Indicators-Actors-etc... collections. One of the more powerful features of systems like ES & Splunk is site-wide indexes - or indexes as a top-level data structure, rather than as a compartment-level data structure. They're basically decoupled from the database organization.

IMO ES and Splunk are both "loosely" organized data pools - which can lead to some unique data organization, searching, and indexing strategies that aren't possible or at least very taxing in more compartmentalized stores like MongoDB or MySQL.

Under the hood, ElasticSearch (and it's cousin, Splunk) both seem to act as "flattened" data pools, where you dictate data grouping through what shares indexes.

So... you could have a site-wide index (to search just common fields on all TLOs) as well as (I think) narrower indexes that only reference fields in specific documents you decide to define as specific TLO's and then only add to a TLO-specific index.

**Comment by mgoffin** [comment]
@ckane I've read up on indexes and types before, and that blog post echoes my concerns:

---
- Fields need to be consistent across types. For instance if two fields have the same name in different types of the same index, they need to be of the same field type (string, date, etc.) and have the same configuration.
- Fields that exist in one type will also consume resources for documents of types where this field does not exist. This is a general issue with Lucene indices: they don’t like sparsity. Sparse postings lists can’t be compressed efficiently because of high deltas between consecutive matches. And the issue is even worse with doc values: for speed reasons, doc values often reserve a fixed amount of disk space for every document, so that values can be addressed efficiently. This means that if Lucene establishes that it needs one byte to store all value of a given numeric field, it will also consume one byte for documents that don’t have a value for this field. Future versions of Elasticsearch will have improvements in this area but I would still advise you to model your data in a way that will limit sparsity as much as possible.
- Scores use index-wide statistics, so scores of documents in one type can be impacted by documents from other types.

This means types can be helpful, but only if all types from a given index have mappings that are similar. Otherwise, the fact that fields also consume resources in documents where they don’t exist could make things worse than if the data had been stored in separate indices.

---

Obviously we just need to strike the right balance, however we need to ensure that whatever we choose can be implemented in any database we decide to support. For example, if we decided to go with a single index in ES but maintain the collections in MongoDB, it would be easy for ES to just ignore the `COL_` settings in settings.py, but if we start mixing and matching, we now need to have mappings built (probably in the backend support for each DB instead of in settings.py).

**Comment by dsnellgrove** [comment]
I'd love to see CRITs use the built-in django support for RDBMS. SQL databases such as Postgres can most definitely scale to CRITs. For example Instagram is Django + postgres. I don't see CRITs having a larger dataset :).  Postgres additionally supports non-relational data types such as a native json storage type. Django provides very powerful tools built on top of this that could signficantly reduce the effort and code involved in maintaining and upgrading models. Additionally our database has continually become more and more a relational database built on top of a non-relational database engine (see: analysis collection, relationships between objects, etc). What benefit does implementing all of CRITs on top of a non-relational database get us that makes it worth the extensive amount of effort required to roll our own model implementation?

**Comment by mgoffin** [comment]
In essence, CRITs wouldn't be tied to Django very much short of its ability to abstract web stuffs. Not sure using Django for db-stuffs is really what we want?

**Comment by dsnellgrove** [comment]
You described declaring your DB type in settings then using a common model that uses the appropriate db-specific magic under the hood. This describes Django Models https://docs.djangoproject.com/en/1.10/topics/db/models/. These have even been expanded to support non-relational databases such as MongoDB before https://github.com/django-nonrel but that work appears to have lost momentum like a lot of other mongodb related libraries. What benefits does CRITs get from using a non-relational database that makes it worth the effort of implementing support for them instead of using the tools already provided by the framework?

**Comment by mgoffin** [comment]
We already use Tastypie for the API. And the biggest complaint about the API is that it is horrifically slow for anything larger than a simple task.

We are providing a way to develop backend support such that if you want to use PostgreSQL, you can write a backend for it that supports all of the database interaction and provides the data to developers in an identical fashion as any other backend would provide so that class methods to work with the data will work with everything.

Django Models might be nice but we are then once again signing ourselves up for someone else's support of features that may or may not be what we want and may or may not have drawbacks that we'll have to deal with (yet again).

**Comment by ckane** [comment]
Though off-topic, I do agree with @dsnellgrove above strongly. As we use CRITs more and more, we are frequently running into hurdles that would be easier to overcome were the back-end built atop some sort of RDBMS. In some cases (campaigns, for instance) it seems like a localized-foreign-value has been pulled into a document, while in other cases (such as comments, or some other related TLOs) a GUID pointer has been brought in instead. In a number of cases, I find myself juggling a lot of query results back and forth in order to handle first, second, etc... order interconnections between objects in order to build out my analysis.

**Comment by mgoffin** [comment]
The only issue I have with using someone else's RDBMS is we are stuck with what they support and the decisions they make. We did that with MongoEngine and they then ripped the Django piece out and screwed us pretty good.

I would be fine with developing our own light-weight RDBMS in a way that all developers are happy as long as it doesn't rely on any 3rd-party libraries and is easier to work with (see the stuff @armtash is doing in #853 and the benchmarks he's getting using pymongo directly).

I spoke with our ES SME at work and he's not as optimistic as I am that making an ES backend is going to be fairly simple (dealing with indexes vs aliases, shards with fixed max documents, mapping files for custom templates, etc.).

What sort of things would y'all want to add to my suggestions/code in the OP that would make this more RDBMS-feeling?

**Comment by apolkosnik** [comment]
Perhaps, it would be great to do the PoC for the Django-native db backends on pgsql, and perhaps even on sqlite, it should not be too hard without ripping out the plumbing.

**Comment by webstergd** [comment]
I would like to second Mike in his suggestion to stripping out the the DB
interaction. We did this in our system and it was a major boon. It made our
lives easier when we needed to adjust the scheme, scale, and incorporate
new DB backends or features. All in all, it greatly increase performance
and decreased our work load. We also didn't have to keep tabs on 3rd party
adjustments. The only real negatives we encountered was in handling race
conditions when supporting simultaneous databases. (I dont think you will
need to worry about this at all). We also had an increase in our upfront
time in having to study how to optimized each database through scheme
designs and perform reads/writes.

Few thoughts:
Their are pros and cons to implementing a database engine as a class or
separate program. A separate program allows you to scale horizontally and
more easily support complex queries through say Spark or Tensorflow. The
wrapper is simpler to work with and decreases the complexity when setting
up the system. Depends on your target audience in my opinion.

Tastypie has a flaw in uploading that causes it to be exceedingly slow.
Only solution is to modify the Tastypie source. Even then we were able to
only get it down to about 1 upload a second. ... drop tastypie

In our experience, treat your search platforms, ES, as a side piece and
separate from the DB backends. Then just have your wrapper or client
provide an abstraction. Makes life exceedingly easier as you don't have to
worry about a generic scheme. Allows you to also support SASI and Solr
easily as well. ... Don't lose hope on ES Mike. It aint that bad.

P.S. I had a kid last week...I am exhausted. I hope this is understandable
and useful.

**Comment by apolkosnik** [comment]
Has anyone looked into django-haystack?

**Comment by dsnellgrove** [comment]
@webstergd That's a lot of useful information/experience, thank you for sharing that. I'm especially interested in the idea of keeping the search platform separate from the DB backend and gluing them together further up the stack. That makes the use cases for each narrower and much easier to implement...

@apolkosnik A lift & shift is probably not that hard here for a proof of concept. Mongo, in their commercial edition, offers an SQL interface to the database. This is actually built on postgres with a foreign data wrapper that connects to the MongoDB.

@mgoffin With your own framework in the mix and an interest in pushing the UI to the client-side instead of integrated in the webserver framework, would we even need Django any more?

**Comment by mgoffin** [comment]
The backend would basically just be a REST framework so we'd still need a web server to host it. Doesn't have to be Django, but would need to be something that can spin up a python backend. Django seems easy, having runserver, but something else would be fine as well. Any preferences?

**Comment by dsnellgrove** [comment]
Easy python webserver with lots of pluggable power? I usually reach for Flask for that kind of work. Django is when you need the whole stack for all the things.

**Comment by webstergd** [comment]
For a very simple and very fast webserver, I like tornado (python) and
golang. Flask is also awesome and you cannot go wrong there either.

**Comment by mgoffin** [comment]
Agreed, Flask is great. But as a production-quality server for the REST framework? Never put Flask through heavy load.

I also don't know how we'd hook auth for the API. I guess a script to generate API keys for accounts that you add through a script as well? (Moving away from Django we lose the management commands).

**Comment by webstergd** [comment]
I have never encountered issues with Flask under large loads. Granted I
switch to Tornado or use golang for the real heavy stuff. Golang handles k
of queries a second without even flinching.

**Comment by amcgregor** [comment]
Howdy!  Sorry for poking my head into a project I'm not actively involved with, but I couldn't help notice the MongoEngine ticket reference.  As some shameless self-promotion (and due to my frustrations with MongoEngine) I have written and recently released my own (explicitly not active record) light-weight object document mapper designed to augment, not replace, the PyMongo driver: [Marrow Mongo](https://github.com/marrow/mongo#readme).  The current query syntax is heavily SQLAlchemy inspired (which I can also _highly_ recommend for the portable RDBMS approach) and support is partially complete (in a PR) for Django-like (or MongoEngine-like) querying.  It's fully tested with code designed for readability: extensive docstrings and comments.  Compatible with 2.7+ and 3.2+ (including Pypy/Pypy3), etc., etc.  💨 

As I'm also noticing discussion of possible microframework choice, I'd also like to point out my [WebCore modular nanoframework](https://github.com/marrow/WebCore#readme) and [web.dispatch.resource](https://github.com/marrow/web.dispatch.resource#readme) RESTful dispatcher.  (We also have a template engine, cinje, offering streaming generation as a directly importable Python DSL. ;)  PyMongo, Marrow Mongo, and SQLAlchemy support are available.  I'd be happy to answer any questions here or on Freenode IRC (#webcore); or, tell me to shove off and I'll remove my comment from here.  ^_^;

Have a great day, and good luck with your architectural decision making!

**Comment by mgoffin** [comment]
@amcgregor Thanks for the update! That's a lot of neat (and relevant!) stuff to look at. Interested in that PR for the Django-like querying as it would make lives easier keeping the querying "as-is".

**Comment by amcgregor** [comment]
@mgoffin [Your wish is my command.](https://github.com/marrow/mongo/releases/tag/1.1.0) — Referred to as "parametric helpers" in [the WIP documentation](https://mongo.webcore.io/).
